File: src/main/java/top/chiloven/lukosbot2/commands/wikis/WikiCommand.java
package top.chiloven.lukosbot2.commands.wikis;

import top.chiloven.lukosbot2.core.command.CommandSource;
import top.chiloven.lukosbot2.model.Attachment;
import top.chiloven.lukosbot2.model.MessageOut;
import top.chiloven.lukosbot2.util.feature.WebScreenshot;
import top.chiloven.lukosbot2.util.feature.WebToMarkdown;
import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.StringArgumentType;
import lombok.extern.log4j.Log4j2;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.stereotype.Service;

import static top.chiloven.lukosbot2.util.brigadier.builder.LiteralArgumentBuilder.literal;
import static top.chiloven.lukosbot2.util.brigadier.builder.RequiredArgumentBuilder.argument;
import static com.mojang.brigadier.arguments.StringArgumentType.greedyString;

/**
 * The /wiki command for Wikipedia screenshots and markdown conversion.
 *
 * @author Chiloven945
 */
@Service
@ConditionalOnProperty(
        prefix = "lukos.commands.control",
        name = "wiki",
        havingValue = "true",
        matchIfMissing = true
)
@Log4j2
public class WikiCommand implements WikiishCommand {

    @Override
    public String defaultLang() {
        return "zh";
    }

    @Override
    public String pathPrefix() {
        return "/wiki/";
    }

    @Override
    public String domainRoot() {
        return "wikipedia.org";
    }

    @Override
    public String name() {
        return "wiki";
    }

    @Override
    public String description() {
        return "维基百科工具，支持截图和转 Markdown";
    }

    @Override
    public String usage() {
        return """
                用法：
                `/wiki <wikipedia_link|article>`    # 生成页面截图
                `/wiki md <wikipedia_link|article>` # 抓取并转为 Markdown 文件
                示例：
                /wiki Java
                /wiki md https://zh.wikipedia.org/wiki/Java
                """;
    }

    @Override
    public void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
                literal(name())
                        // /wiki <link>
                        .then(argument("link", greedyString())
                                .executes(ctx -> {
                                    String link = StringArgumentType.getString(ctx, "link").trim();
                                    runScreenshot(ctx.getSource(), link);
                                    return 1;
                                })
                        )
                        // /wiki md <link>
                        .then(literal("md")
                                .then(argument("link", greedyString())
                                        .executes(ctx -> {
                                            String link = StringArgumentType.getString(ctx, "link").trim();
                                            runMarkdown(ctx.getSource(), link);
                                            return 1;
                                        })
                                )
                        )
                        // No args
                        .executes(ctx -> {
                            ctx.getSource().reply(usage());
                            return 1;
                        })
        );
    }

    private void runScreenshot(CommandSource src, String link) {
        try {
            String url = normalize(link);
            if (isNot(url)) {
                src.reply("仅支持 Wikipedia 链接。示例：/wiki https://en.wikipedia.org/wiki/Java");
            }
            var img = WebScreenshot.screenshotWikipedia(url);
            MessageOut out = MessageOut.text(src.in().addr(), null)
                    .with(Attachment.imageBytes(img.filename(), img.bytes(), img.mime()));
            src.reply(out);
        } catch (Exception e) {
            log.warn("Wiki screenshot failed: {}", link, e);
            src.reply("截图失败：" + e.getMessage());
        }
    }


    private void runMarkdown(CommandSource src, String link) {
        try {
            String url = normalize(link);
            if (isNot(url)) {
                src.reply("仅支持 Wikipedia 链接。示例：/wiki md https://zh.wikipedia.org/wiki/Java");
            }
            var md = WebToMarkdown.fetchWikipediaMarkdown(url);
            MessageOut out = MessageOut.text(src.in().addr(), "已转换为 Markdown")
                    .with(Attachment.fileBytes(md.filename(), md.bytes(), md.mime()));
            src.reply(out);
        } catch (Exception e) {
            log.warn("Wiki md failed: {}", link, e);
            src.reply("转换失败：" + e.getMessage());
        }
    }

}

File: src/main/java/top/chiloven/lukosbot2/commands/bot/wikis/WikiCommand.kt
package top.chiloven.lukosbot2.commands.bot.wikis

import org.apache.logging.log4j.LogManager
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
import org.springframework.stereotype.Service
import top.chiloven.lukosbot2.commands.IBotCommand
import top.chiloven.lukosbot2.core.command.bot.CommandSource
import top.chiloven.lukosbot2.core.command.definition.dsl.arg
import top.chiloven.lukosbot2.core.command.definition.dsl.botCommand
import top.chiloven.lukosbot2.core.command.definition.dsl.oneOf
import top.chiloven.lukosbot2.core.model.message.media.BytesRef
import top.chiloven.lukosbot2.core.model.message.outbound.OutFile
import top.chiloven.lukosbot2.core.model.message.outbound.OutImage
import top.chiloven.lukosbot2.core.model.message.outbound.OutboundMessage
import top.chiloven.lukosbot2.util.feature.WebScreenshot
import top.chiloven.lukosbot2.util.feature.WebToMarkdown

@Service
@ConditionalOnProperty(
    prefix = "lukos.commands.control",
    name = ["wiki"],
    havingValue = "true",
    matchIfMissing = true
)
class WikiCommand : IBotCommand, IWikiishCommand {

    private val log = LogManager.getLogger(WikiCommand::class.java)

    override fun defaultLang() = "zh"
    override fun pathPrefix() = "/wiki/"
    override fun domainRoot() = "wikipedia.org"

    override fun definition() = botCommand("wiki") {
        description = "查询维基百科，支持截图和导出 Markdown"

        literal("md") {
            description = "抓取整页并转为 Markdown 文件"
            raw("link") { link ->
                runMarkdown(source, link.trim())
            }
            param("link", "维基百科链接")
            param("article", "词条名（可用 `en:` 前缀指定语言）")
        }

        raw("link", required = false) { link ->
            if (link.isBlank()) sendUsage(source)
            else runScreenshot(source, link.trim())
        }

        syntax(
            "生成页面截图",
            oneOf(
                arg("link"),
                arg("article")
            )
        )

        example(
            "wiki Java",
            "wiki en:Albert_Einstein",
            "wiki md https://zh.wikipedia.org/wiki/Java"
        )
        note("词条名支持 `en:` 前缀指定语言，例如：`en:Albert_Einstein`。")
    }

    private fun runScreenshot(src: CommandSource, linkOrTitle: String) {
        try {
            val url = normalize(linkOrTitle)
            if (isNot(url)) {
                src.reply("仅支持 Wikipedia 链接或词条名。")
                return
            }

            val img = WebScreenshot.screenshotWikipedia(url)
            val ref = BytesRef(img.filename(), img.bytes(), img.mime())
            src.reply(OutboundMessage(src.addr(), listOf(OutImage(ref, null, img.filename(), img.mime()))))
        } catch (e: Exception) {
            log.warn("Wiki screenshot failed: {}", linkOrTitle, e)
            src.reply("截图失败：${e.message}")
        }
    }

    private fun runMarkdown(src: CommandSource, linkOrTitle: String) {
        try {
            val url = normalize(linkOrTitle)
            if (isNot(url)) {
                src.reply("仅支持 Wikipedia 链接或词条名。")
                return
            }

            val md = WebToMarkdown.fetchWikipediaMarkdown(url)
            val ref = BytesRef(md.filename(), md.bytes(), md.mime())
            src.reply(
                OutboundMessage(
                    src.addr(),
                    listOf(
                        OutFile(
                            ref,
                            "已转换为 Markdown。",
                            md.filename(),
                            md.mime()
                        )
                    )
                )
            )
        } catch (e: Exception) {
            log.warn("Wiki md failed: {}", linkOrTitle, e)
            src.reply("转换失败：${e.message}")
        }
    }

}
